src/adaptors/http1/http1_adaptor.c

```diff
@@ -228,59 +228,116 @@ void qdr_http1_error_response(qdr_http1_request_base_t *hreq,
 // Raw Connection Write Buffer Management
 //
 
-
-// Write list of data out the raw connection, freeing entries when data is exhausted
-//
-uint64_t qdr_http1_write_out_data(qdr_http1_connection_t *hconn, qdr_http1_out_data_list_t *fifo)
+static uint32_t qdr_http1_get_out_buffers(qdr_http1_out_data_list_t *fifo,
+                                          qd_adaptor_buffer_list_t  *abuf_list,
+                                          size_t                     limit)
 {
-    size_t count = !hconn->raw_conn || pn_raw_connection_is_write_closed(hconn->raw_conn)
-        ? 0
-        : pn_raw_connection_write_buffers_capacity(hconn->raw_conn);
+    qd_adaptor_buffer_t  *abuf         = 0;
+    qdr_http1_out_data_t *od           = 0;
+    uint32_t              total_octets = 0;
+    while (!DEQ_IS_EMPTY(*fifo) && limit > 0) {
+        if (!od)
+            od = DEQ_HEAD(*fifo);
 
-    if (hconn->write_buf_busy || count == 0)
-        return 0;
+        if (!abuf)
+            abuf = qd_adaptor_buffer();
+
+        size_t data_octets = qd_iterator_remaining(od->data_iter);
+        assert(data_octets > 0);
 
-    const size_t max_octets = HTTP1_IO_BUF_SIZE;
-    size_t total_octets = 0;
-    while (!DEQ_IS_EMPTY(*fifo) && total_octets < max_octets) {
-        qdr_http1_out_data_t *od = DEQ_HEAD(*fifo);
-        uint32_t data_octets = qd_iterator_remaining(od->data_iter);
+        unsigned char *abuf_cursor   = qd_adaptor_buffer_cursor(abuf);
+        size_t         abuf_capacity = qd_adaptor_buffer_capacity(abuf);
+        assert(abuf_capacity > 0);
 
-        size_t len = MIN(data_octets, max_octets - total_octets);
-        int copied = qd_iterator_ncopy(od->data_iter, &hconn->write_buffer[total_octets], len);
-        assert(copied == len);
-        data_octets -= copied;
+        size_t copied = qd_iterator_ncopy_octets(od->data_iter, abuf_cursor, MIN(data_octets, abuf_capacity));
+        assert(copied > 0);
         total_octets += copied;
 
-        qd_iterator_trim_view(od->data_iter, data_octets);
         if (qd_iterator_remaining(od->data_iter) == 0) {
             DEQ_REMOVE_HEAD(*fifo);
             _free_qdr_http1_out_data(od);
+            od = 0;
+        }
+
+        qd_adaptor_buffer_insert(abuf, copied);
+        if (qd_adaptor_buffer_capacity(abuf) == 0) {
+            DEQ_INSERT_TAIL(*abuf_list, abuf);
+            abuf = 0;
+            limit -= 1;
         }
     }
 
-    if (total_octets) {
-        pn_raw_buffer_t pn_buff = {0};
-        pn_buff.bytes = (char*) hconn->write_buffer;
-        pn_buff.size  = total_octets;
+    if (abuf) {
+        DEQ_INSERT_TAIL(*abuf_list, abuf);
+    }
+    return total_octets;
+}
+
+// Give a list of output buffers to the raw connection for sending. The raw connection will take as many buffers off the
+// list as it has capacity for. Return the number of buffers taken.
+//
+static size_t qdr_http1_write_buffers(pn_raw_connection_t *raw_conn, qd_adaptor_buffer_list_t *abuf_list)
```

Comment:
Can this function be replaced with a call to https://github.com/skupperproject/skupper-router/blob/main/src/adaptors/adaptor_common.c#L256

Reply:
Done